enhancements/ingress/optional-ingress.md

```diff
@@ -0,0 +1,274 @@
+No change to the capability API is needed, a simple `if` can ensure this pre-requisite.
+
+#### Ingress operator APIs
+
+The following APIs are not available if the ingress capability is disabled:
+- CRDs which are part of the cluster operator's manifest payload:
+  - [IngressController](https://github.com/openshift/cluster-ingress-operator/blob/master/manifests/00-custom-resource-definition.yaml)
+  - [DNSRecord](https://github.com/openshift/cluster-ingress-operator/blob/master/manifests/00-custom-resource-definition-internal.yaml)
+- CRDs which are created at runtime if `gateway-api` feature gate is set:
+  - [GatewayClass](https://github.com/openshift/cluster-ingress-operator/blob/master/assets/gateway-api/gateway.networking.k8s.io_gatewayclasses.yaml)
+  - [Gateway](https://github.com/openshift/cluster-ingress-operator/blob/master/assets/gateway-api/gateway.networking.k8s.io_gateways.yaml)
+  - [HTTPRoute](https://github.com/openshift/cluster-ingress-operator/blob/master/assets/gateway-api/gateway.networking.k8s.io_httproutes.yaml)
+
+#### Related APIs
+
+The following APIs are left intact if the ingress capability is disabled:
+- [Route API](https://docs.openshift.com/container-platform/4.13/rest_api/network_apis/route-route-openshift-io-v1.html). Implemented in two places: OpenShift API as [CRD](https://github.com/openshift/api/blob/master/route/v1/route.crd.yaml) and in [openshift-api-server](https://github.com/openshift/openshift-apiserver/tree/master/pkg/route).
+- [Ingress Config API](https://docs.openshift.com/container-platform/4.13/rest_api/config_apis/ingress-config-openshift-io-v1.html). Implemented in OpenShift API as [CRD](https://github.com/openshift/api/blob/master/config/v1/0000_10_config-operator_01_ingress.crd.yaml).
+
+However the implementation of these APIs (or parts of them) depend on the ingress operator:
+- The ingress operator creates the default ingress controller to manage the backends for the routes.
+- For [the component routes](https://docs.openshift.com/container-platform/4.13/rest_api/config_apis/ingress-config-openshift-io-v1.html#spec-componentroutes) the ingress operator creates the required role and role bindings to get the custom TLS certificate secret from `openshift-config` namespace.
+
+#### Console operator
+
+To prevent the console operator from going degraded, [the console route check](https://github.com/openshift/console-operator/blob/master/pkg/console/controllers/healthcheck/controller.go#L122) should be skipped if the ingress capability is not enabled in `ClusterVersion`.
```

Comment:
Should it?  The Route API will still exist.  Perhaps the API will be delivered via another controller chain (e.g. including the AWS load-balancer operator).  If folks ask for the `Console` component but do not provide a controller that implements the Route, that seems like it would be worth the console operator complaining about.

Reply:
I didn't see any ingress controller which implements the Route API. The AWS LB controller while being the preferred solution for ROSA users doesn't work with any APIs rather that Kube's. However I agree that theoretically this can happen, for instance, we can implement the Route API support in the OpenShift fork of the AWS LB controller. Although I would doubt that this is something that we would like to add to the scope of this enhancement.
So, practically speaking, as of now, we have a choice of letting the controller operator go degraded or try to tolerate this. I think that from the point of view of the customer who requested the change, more logical would be to not see any operator going degraded when this feature is delivered.

Reply:
> I didn't see any ingress controller which implements the Route API.

The ingress operator [requests `openshift-ingress-router` RBAC to monitor Routes and update their `status`][1].  And checking audit logs for [4.14 CI][2]:

```console
$ curl -s https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/logs/periodic-ci-openshift-release-master-nightly-4.14-e2e-aws-sdn-serial/1661746644430884864/artifacts/e2e-aws-sdn-serial/gather-audit-logs/artifacts/audit-logs.tar | tar -xz --strip-components=2
$ zgrep -h '"resource":"routes"' kube-apiserver/*.log.gz | jq -r 'select(.verb == "update" and .objectRef.subresource == "status") | .user.username' | sort | uniq -c
     17 system:serviceaccount:openshift-ingress:router
```

I'm not entirely clear on where that code lives, but it seems like something that would be disabled by the `Ingress` component.

> However I agree that theoretically this can happen, for instance, we can implement the Route API support in the OpenShift fork of the AWS LB controller.

That's one option, but I agree that it's heavy.  Another option would be creating a shim Route controller that watched for Routes, translated them into Ingress requests for the AWS load balancer component to manage, and then proxied the `status` back from the Ingress resource to the Route resource.  I have no idea how close the types are; maybe that sort of proxying is impossible, without knowledge of the specific Service being exposed?

[1]: https://github.com/openshift/cluster-ingress-operator/blob/e068d04e01cca70be8efb0a595188835525b4792/assets/router/cluster-role.yaml#L31-L44
[2]: https://prow.ci.openshift.org/view/gs/origin-ci-test/logs/periodic-ci-openshift-release-master-nightly-4.14-e2e-aws-sdn-serial/1661746644430884864

Reply:
> > I didn't see any ingress controller which implements the Route API.
> 
> The ingress operator [requests openshift-ingress-router RBAC to monitor Routes and update their status](https://github.com/openshift/cluster-ingress-operator/blob/e068d04e01cca70be8efb0a595188835525b4792/assets/router/cluster-role.yaml#L31-L44). And checking audit logs for [4.14 CI](https://prow.ci.openshift.org/view/gs/origin-ci-test/logs/periodic-ci-openshift-release-master-nightly-4.14-e2e-aws-sdn-serial/1661746644430884864):
> 
> $ curl -s https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/logs/periodic-ci-openshift-release-master-nightly-4.14-e2e-aws-sdn-serial/1661746644430884864/artifacts/e2e-aws-sdn-serial/gather-audit-logs/artifacts/audit-logs.tar | tar -xz --strip-components=2
> $ zgrep -h '"resource":"routes"' kube-apiserver/*.log.gz | jq -r 'select(.verb == "update" and .objectRef.subresource == "status") | .user.username' | sort | uniq -c
>      17 system:serviceaccount:openshift-ingress:router
> I'm not entirely clear on where that code lives, but it seems like something that would be disabled by the Ingress component.

Sorry, I meant that I didn't see any ingress controller apart from the OpenShift ingress controller. Something like nginx-ingress-controller or kong-ingress-controller which could replace the ingerss controller spawned by the cluster ingress operator.

> > However I agree that theoretically this can happen, for instance, we can implement the Route API support in the OpenShift fork of the AWS LB controller.
> 
> That's one option, but I agree that it's heavy. Another option would be creating a shim Route controller that watched for Routes, translated them into Ingress requests for the AWS load balancer component to manage, and then proxied the status back from the Ingress resource to the Route resource. I have no idea how close the types are; maybe that sort of proxying is impossible, without knowledge of the specific Service being exposed?

A controller which translates from routes to ingresses, this goes in the opposite direction to [the "ingress to route" translation](https://github.com/openshift/route-controller-manager#ingress-to-route-controller) we already have in place.

Reply:
@wking, are you proposing that enabling the "Console" component should require enabling the "Ingress" component?

Reply:
>@wking, are you proposing that enabling the "Console" component should require enabling the "Ingress" component?

@Miciah: as of now we decided to not make that kind of dependency between the console and the ingress operators. The only change will be in the console operator to tolerate the absence of the ingress operator (`Console Operator` chapter).